### PR TITLE
Center rewards coefficient [skip-e2e]

### PR DESCRIPTION
**Pull request by yardenhoch**
# Add center_rewards_coefficient support for Reward Model training

## Description

This PR adds support for the `center_rewards_coefficient` parameter in Axolotl's reward model training, enabling users to incentivize mean-zero reward outputs for improved training stability. The parameter adds an auxiliary loss term that encourages the reward model to produce rewards centered around zero, eliminating the need for post-training normalization.

**Key Changes:**
- Added `center_rewards_coefficient` to Pydantic schema configuration
- Integrated parameter flow through training argument builders  
- Added documentation with usage examples
- Updated example configurations to demonstrate the feature
- Comprehensive test coverage for parameter integration

## Motivation and Context

The `center_rewards_coefficient` parameter implements the reward centering technique proposed in [Eisenstein et al., 2023](https://huggingface.co/papers/2312.09244) (Equation 2). This technique helps ensure reward models naturally produce mean-zero outputs by adding an auxiliary loss:

```
L_aux = center_rewards_coefficient × (sum of rewards)²
L_total = L_reward + L_aux
```

**Benefits:**
- Improved training stability for reward models
- Eliminates need for post-training reward normalization
- Better alignment with RLHF best practices
- Seamless integration with existing TRL infrastructure

**Problem Solved:** Without reward centering, reward models can drift toward consistently positive or negative outputs, requiring manual normalization and potentially affecting downstream RLHF training quality.

### Local Testing
- **Unit Tests**: Comprehensive test suite in `tests/test_center_rewards_coefficient.py` covering:
  - Pydantic schema integration and validation
  - Parameter flow through builder classes
  - Default behavior and edge cases
  - Complete flow verification from config to TRL RewardConfig

### Integration Points Verified
1. **Schema Integration** ✅: Parameter properly validates in `AxolotlInputConfig`
2. **Builder Integration** ✅: Parameter flows through `TrainerBuilderBase._set_base_training_args`  
3. **Training Args** ✅: Parameter reaches `AxolotlRewardConfig` and TRL's `RewardConfig`
4. **Documentation** ✅: Clear examples and usage patterns

### Test Environment
- **Local**: Python 3.12.6, macOS (schema and builder logic validation)
- **Production**: Planned testing via Argo workflow with full training environment

### Example Usage Validation
```yaml
reward_model: true
center_rewards_coefficient: 0.01  # Recommended value
datasets:
  - path: argilla/distilabel-intel-orca-dpo-pairs
    type: bradley_terry.chat_template
```

## Types of changes

- [x] **New feature** (non-breaking change which adds functionality)
- [x] **Documentation update** (adds usage examples and explanation)
- [x] **Test coverage** (comprehensive unit tests for the new feature)

**Files Modified:**
- `src/axolotl/utils/schemas/config.py` - Added parameter to Pydantic schema
- `src/axolotl/core/builders/base.py` - Added to training args passthrough
- `src/axolotl/core/builders/causal.py` - Explicit reward model integration
- `docs/reward_modelling.qmd` - Added documentation and examples
- `examples/gemma2/reward-model.yaml` - Updated example config
- `examples/qwen2/reward-model.yaml` - Updated example config  
- `tests/test_center_rewards_coefficient.py` - Comprehensive test suite

## Summary by CodeRabbit

* New Features
  * Added support for a center_rewards_coefficient option in reward-model training to encourage mean-zero rewards and improve stability.
  * Updated example configurations to include center_rewards_coefficient (default example value: 0.01).

* Documentation
  * Added a “Centering Rewards” section explaining the concept, recommended values, configuration usage, and reference.

* Tests
  * Added integration tests covering acceptance, default behavior, and propagation of center_rewards_coefficient into training arguments.

### CODE REVIEW

**Comment by salmanmohammadi** [comment]
Thanks so much for this PR @yardenhoch. This looks great to me. A few small nits:
- I think the unit test might be overkill for this use-case, let's remove it.
- The default behaviour in TRL is to disable this feature and we should follow suit. You can highlight that we support this by adding it as a commented line in our reward model configs though - that would be great.
- Lastly, having an entire section in the docs for this feature may also be overkill. A sentence under the "(Outcome) Reward Models" section would be sufficient - with a link to the corresponding note in the [TRL docs](https://huggingface.co/docs/trl/v0.22.1/en/reward_trainer#centering-rewards).

Feel free to ping me once you've had a chance to make these changes.

File: docs/reward_modelling.qmd [hunk 1]
```diff
@@ -39,6 +39,29 @@ Bradley-Terry chat templates expect single-turn conversations in the following f
...
 
+#### Centering Rewards
+
+For improved training stability, you can enable reward centering which encourages the model to produce mean-zero outputs by adding an auxiliary loss:
+
+```yaml
+base_model: google/gemma-2-2b
+model_type: AutoModelForSequenceClassification
+num_labels: 1
+tokenizer_type: AutoTokenizer
```
Comment (NanoCode012) [inline docs/reward_modelling.qmd]:
```suggestion
```
Are these needed as is duplicate of above? Maybe we can simplify

File: docs/reward_modelling.qmd [hunk 2]
```diff
@@ -39,6 +39,29 @@ Bradley-Terry chat templates expect single-turn conversations in the following f
...
+
+reward_model: true
+center_rewards_coefficient: 0.01  # Typical values: 0.001 - 0.1
+chat_template: gemma
+datasets:
+  - path: argilla/distilabel-intel-orca-dpo-pairs
+    type: bradley_terry.chat_template
+
+val_set_size: 0.1
+eval_steps: 100
```
Comment (NanoCode012) [inline docs/reward_modelling.qmd]: Similarly, unneeded portion
```suggestion
```

File: src/axolotl/core/builders/base.py [hunk 3]
```diff
@@ -498,6 +498,8 @@ def _set_base_training_args(
             "dion_momentum",
             "dion_rank_fraction",
             "dion_rank_multiple_of",
+            # reward modeling
+            "center_rewards_coefficient",
```
Comment (NanoCode012) [inline src/axolotl/core/builders/base.py]:
```suggestion
```

My bad, we didn't need this. It is a duplicate of setting in `src/axolotl/core/builders/causal.py`.

File: examples/gemma2/reward-model.yaml [hunk 4]
```diff
@@ -7,6 +7,7 @@ tokenizer_type: AutoTokenizer
 # hub_model_id: username/custom_model_name
 
 reward_model: true
+center_rewards_coefficient: 0.01  # Incentivize mean-zero rewards for improved stability
```
Comment (NanoCode012) [inline examples/gemma2/reward-model.yaml]: Instead of setting it as a default in our older configs, would you be interested in making a new one for a newer model that you may have trained recently on?


```suggestion
```

File: tests/test_center_rewards_coefficient.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        cfg = self._get_base_cfg()
+        cfg["center_rewards_coefficient"] = 0.01
+        config = AxolotlInputConfig(**cfg)
+
+        training_args_kwargs = {}
+        for arg in ["center_rewards_coefficient"]:
+            if getattr(config, arg, None) is not None:
+                training_args_kwargs[arg] = getattr(config, arg)
+
+        self.assertIn("center_rewards_coefficient", training_args_kwargs)
```
Comment (NanoCode012) [inline tests/test_center_rewards_coefficient.py]: I'm not sure what's the use case for this test? You may want to refer to our `tests/core/test_builders.py` for how to load the trainer builder.

**Review by NanoCode012 — COMMENTED** [review]
Thanks! I left some comments

**Comment by yardenhoch** [comment]
@NanoCode012 @SalmanMohammadi 
ready to review

File: docs/reward_modelling.qmd [hunk 6]
```diff
@@ -39,6 +39,23 @@ Bradley-Terry chat templates expect single-turn conversations in the following f
 }
 ```
 
+#### Centering Rewards
```
Comment (salmanmohammadi) [inline docs/reward_modelling.qmd]: As I mentioned:
> Lastly, having an entire section in the docs for this feature may also be overkill. A sentence under the "(Outcome) Reward Models" section would be sufficient - with a link to the corresponding note in the [TRL docs](https://huggingface.co/docs/trl/v0.22.1/en/reward_trainer#centering-rewards).

File: tests/test_center_rewards_coefficient.py [hunk 7]
```diff
@@ -0,0 +1,86 @@
+import unittest
```
Comment (salmanmohammadi) [inline tests/test_center_rewards_coefficient.py]: I don't think we need this file.

**Comment by yardenhoch** [comment]
@SalmanMohammadi done

**Review by salmanmohammadi — APPROVED** [review]
Thanks so much for this @yardenhoch!

File: examples/qwen3/reward-model.yaml [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+tf32: true
+
+gradient_checkpointing: true
+gradient_checkpointing_kwargs:
+  use_reentrant: false
+warmup_steps: 10
+logging_steps: 1
+weight_decay: 0.01
+
+save_strategy: "no"
```
Comment (NanoCode012) [inline examples/qwen3/reward-model.yaml]:
```suggestion
```
For a basic config, we'd want saving

File: examples/qwen3/reward-model.yaml [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+use_safetensors: true
+deepspeed: deepspeed_configs/zero1.json
+
+wandb_project:
+wandb_entity:
+wandb_watch:
+wandb_name:
+wandb_log_model:
+
+gradient_accumulation_steps: 16
```
Comment (NanoCode012) [inline examples/qwen3/reward-model.yaml]:
```suggestion
gradient_accumulation_steps: 4
```
This was a very high default. Reducing to a smaller number.

File: examples/qwen3/reward-model.yaml [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+val_set_size: 0.0
+output_dir: ./outputs/out
+remove_unused_columns: false
+
+sequence_len: 8192
+sample_packing: false
+eval_sample_packing: false
+pad_to_sequence_len: true
+
+use_safetensors: true
```
Comment (NanoCode012) [inline examples/qwen3/reward-model.yaml]:
```suggestion
```
This is default, so removing this.

File: examples/qwen3/reward-model.yaml [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+reward_model: true
+center_rewards_coefficient: 0.01  # Incentivize mean-zero rewards for improved stability
+chat_template: qwen3
+datasets:
+  - path: argilla/distilabel-intel-orca-dpo-pairs
+    type: bradley_terry.chat_template
+
+val_set_size: 0.0
+output_dir: ./outputs/out
+remove_unused_columns: false
```
Comment (NanoCode012) [inline examples/qwen3/reward-model.yaml]:
```suggestion
```

Might not be necessary? I could be wrong.

File: examples/qwen3/reward-model.yaml [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+lr_scheduler: linear
+learning_rate: 0.00002
+
+bf16: true
+tf32: true
+
+gradient_checkpointing: true
+gradient_checkpointing_kwargs:
+  use_reentrant: false
+warmup_steps: 10
```
Comment (NanoCode012) [inline examples/qwen3/reward-model.yaml]:
```suggestion
warmup_ratio: 0.1
```
We recommend ratio over steps as it's independent of batch size.

**Review by NanoCode012 — CHANGES_REQUESTED** [review]
Thanks, I just have some minor corrections on the config.

**Review by salmanmohammadi — APPROVED** [review]
